PLG-Motion | Example dataset & Core DB features (#1062)
arangodb/docs-hugo: site/content/contextual-data-platform/tutorials/core.md

---
title: ArangoDB core database features
menuTitle: Core features
weight: 15
description: >-
  Learn the essential ArangoDB features by querying the movie graph: documents,
  AQL, indexes, joins, graph traversals, and vector search
---

> In this tutorial, we learn the core features of ArangoDB by exploring the movie

Reviewer: I will only mention this once here, bit I think this holds true for all tutorials.

In the initial description of the Tutorial we lay out the scope as "WE".
`we learn the core features`

Later in the tutorial it is always "YOU".

Is this intentional, and should we align it to be always WE or always YOU (I have not preference which)?

Author: Should be you, as we is ambiguous

dataset hands-on. We move from looking up a single document to filtering,

> sorting, joining collections, and finally traversing the movie graph — the

Reviewer: ```suggestion
sorting, joining collections, and finally traversing the movie graph, the
```

do not make it to obviously LLM generated ;)

essential skills every ArangoDB user needs.

This is learning-oriented: follow along query by query. We do not aim to cover
everything, only the essentials, using data you can see change as you go. Every
query below runs against the movie dataset from the previous tutorial, so you
can paste each one in and watch the results build on the last.

By the end, you will be able to:

- Navigate the data model: databases, collections, and documents.
- Read, filter, sort, and limit data with AQL.
- Insert, update, and remove a document.
- Speed up a query with an index and read its query explanation.
- Join collections and traverse the graph to answer connected questions.
- Find semantically similar movies with a vector index.

## Prerequisites

- The [example movie dataset](dataset.md) loaded into a `movies` database.

> - Access to the data platform web interface, where you run every query in this

Reviewer: Please add a link to "https://localhost:8529/ui/"

assumes `movies` is the active database.

## Step 1: Explore the data model

ArangoDB organizes data in a simple hierarchy: a **database** contains
**collections**, and a collection contains **documents**. A document is just a
JSON object, and ArangoDB is schema-free — two documents in the same collection
need not have the same fields. Every document carries three system attributes:
`_key` (unique within its collection), `_id` (the collection name and key joined
as `Movies/8961`), and `_rev` (a revision marker ArangoDB manages for you).

There are two kinds of collections. **Document (node) collections** such as
`Movies` and `People` hold the "things". **Edge collections** such as `HasGenre`
and `ActedIn` hold the relationships: every edge is a document that additionally
has a `_from` and a `_to` attribute, each pointing at a document by its `_id`.
Edges are what let us traverse the graph later in this tutorial.

## Step 2: Run your first AQL query

**AQL** (the ArangoDB Query Language) is the single language you use for
everything: reading documents, filtering, joining collections, and traversing
graphs. Every step below is an AQL query you paste into the Query editor.

> Open the **Query editor** in the web interface. It has an input area where you

Reviewer: Please add the Icon for "Query Editor"

> type a query, a **Run** button (or the run shortcut) that executes it, and a

Reviewer: The new one, first asks to click on "New Query" button.

results panel below that shows what came back. To confirm the editor works, run
the simplest possible query:

```aql

> RETURN "Hello ArangoDB"

Reviewer: ```suggestion
RETURN "Hello Arango.ai"
```

```

The results panel shows a single string. `RETURN` is how every AQL query
produces output; the rest of this tutorial is mostly about deciding *what* to
return.

## Let AI write the queries for you

Before we dive deeper into AQL, it is worth knowing that the data platform can
write queries *for* you, right inside the Query editor:

- **AQLizer** turns a plain-language request into an AQL query. Click the

>   **AQLizer** button, describe what you want (for example, "list the 10

Reviewer: Worth mentioning that you will need an OpenAI API key.

  highest-rated movies with at least 1000 votes"), and it generates the query,
  which you open in a tab to run. This can make it unnecessary to write AQL by
  hand for everyday questions. It is a data platform feature — see
  [Generate queries (AQLizer)](../../platform-suite/query-editor.md#generate-queries-aqlizer)
  and the [Natural Language to AQL](../../agentic-ai-suite/natural-language-to-aql/_index.md)
  introduction.

> - **Reasoner** analyzes an existing AQL query and returns a validated, faster

Reviewer: I would split the Reasoner out here.
I would create a badly performing query later on, and then use the Reasoner as an example for optimizing it.

  version with a performance comparison. Click **Optimize** to use it. See
  [Optimize queries (Reasoner)](../../platform-suite/query-editor.md#optimize-queries-reasoner).

So why learn AQL at all? Because generation is a starting point, not the finish
line: AI can make mistakes, so you always verify what it produces, and real work
means reading a generated query, correcting it, and extending it into something
more specific. The rest of this tutorial gives you exactly that foundation — with
it, the generated queries become something you can trust, adapt, and build on

> rather than a black box.

Reviewer: Let us try to formulate this a bit differently.
Clients are looking at this to avoid learning AQL.
We should not point them on "does not work" right away.

Let us try to formulate in a way:
That this gives you skill to overcome limits of LLM query generation and allows you to judge and understand the query the LLM has generated.

## Step 3: Read documents

The most direct read is a lookup by document ID with the `DOCUMENT()` function.
This fetches exactly one document (Bad Boys II):

```aql

> RETURN DOCUMENT("Movies/8961")

Reviewer: I would skip DOCUMENT function.
Let us focus on real AQL.

Looking things up by "random" ids is nothing I would really want to see in a tutorial.

If the key would have some meaning (e.g. the actor name) than I would be okay with it.

Author: We could modify the dataset but I think this is closer to reality and DOCUMENT() tends to perform better for a small number of documents than a FOR ... FILTER. There are use cases for it, but I agree that it's probably not needed in this tutorial.

```

More often you iterate over a collection with a `FOR` loop. `LIMIT` caps how many
documents you get back, which keeps exploratory queries fast on a 45,000-movie
collection:

```aql
FOR m IN Movies
  LIMIT 5
  RETURN m
```

> Returning the whole document is rarely what you want. Build a new object in the

Reviewer: ```suggestion
Typically in the use-cases you only have a need for very specific attributes, hence you can build a new object in the
```

```aql
FOR m IN Movies
  LIMIT 5
  RETURN { title: m.title, year: m.release_date }
```

Here `m` is the loop variable — one movie document per iteration — and `m.title`
reads a single attribute from it.

## Step 4: Filter documents

Add a `FILTER` to keep only the documents that match a condition. Read it as
"find movies where…":

```aql
FOR m IN Movies
  FILTER m.vote_average > 8
  RETURN m.title
```

Combine conditions with `AND` and `OR`, and use range comparisons to narrow a
result set. This finds well-rated movies released in this century that enough
people actually voted on:

```aql
FOR m IN Movies
  FILTER m.vote_average >= 8
     AND m.vote_count >= 1000
     AND m.release_date >= "2000-01-01"
  RETURN { title: m.title, rating: m.vote_average }
```

Because `release_date` is stored as an ISO date string (`"2003-07-18"`), string
comparison and date comparison agree, so `>= "2000-01-01"` does what you expect.

## Step 5: Sort and limit

`SORT` orders the results; combine it with `LIMIT` to get a top-N. This is the
classic "top 10 highest-rated movies" query:

```aql
FOR m IN Movies
  SORT m.vote_average DESC
  LIMIT 10
  RETURN { title: m.title, rating: m.vote_average }
```

> Sort by more than one attribute to break ties — here, when two movies have the
> same average, the one with more votes ranks higher:
> 
> ```aql
> FOR m IN Movies
>   SORT m.vote_average DESC, m.vote_count DESC
>   LIMIT 10
>   RETURN { title: m.title, rating: m.vote_average, votes: m.vote_count }
> ```

Reviewer: ```suggestion
```

This gets too much into details for the tutorial.
It is good to know, but the goal here is not to make an AQL expert in the end

`FILTER`, `SORT`, and `LIMIT` compose naturally. Put them together for
"top-rated recent movies with a meaningful number of votes":

```aql
FOR m IN Movies
  FILTER m.release_date >= "2010-01-01" AND m.vote_count >= 1000
  SORT m.vote_average DESC
  LIMIT 10
  RETURN { title: m.title, rating: m.vote_average, released: m.release_date }
```

## Step 6: Compute and aggregate

You are not limited to returning stored fields — you can compute values in the
query. AQL ships with functions for strings, numbers, dates, arrays, and more.
This uses `LENGTH` (string length) and `LEFT` (the first characters of a string)
to derive new fields:

> 

Reviewer: Skip this example

```aql
FOR m IN Movies
  FILTER m.release_date != null
  LIMIT 5
  RETURN {
    title: m.title,
    title_length: LENGTH(m.title),
    release_year: LEFT(m.release_date, 4)
  }
```

> To summarize many documents into a few numbers, use `COLLECT`. Grouping with

Reviewer: Keep this, it shows how grouping works.
And does some AQL Functions.

I would actually use `DATE_YEAR(m.release_date)` instead of this LEFT computation

execution plan without running the query:

```aql
FOR m IN Movies
  FILTER m.title == "Bad Boys II"
  RETURN m
```

Explain reports a scan over the whole `Movies` collection: ArangoDB has no faster
way to find the matching title.

Now create an index and check again:

1. In the left sidebar of the web interface, go to
  **Management** > **Collections** and open the `Movies` collection.
2. Go to the **Indexes** tab.
3. Add a **persistent index** on the `title` field.
4. Run **Explain** on the same query once more: the plan now uses an index lookup
   instead of a collection scan, because ArangoDB can jump straight to the matching
   document.
   
Indexes speed up filtering and sorting on the indexed attributes, at the cost of
a little extra storage and slightly slower writes — so you index the fields you
query on, not every field.

## Step 9: Join collections

Sometimes the data you want lives in another collection. The `Movies` and
`MovieEmbeddings` collections share the same `_key` for each movie (the
embeddings were kept separate so the movie documents stay lean), so you can join
them by primary key. The efficient way is a direct `DOCUMENT()` lookup using the
movie's own key:

> ```aql
> FOR m IN Movies
>   FILTER m.vote_average >= 8
>   LIMIT 5
>   LET emb = DOCUMENT("MovieEmbeddings", m._key)
>   RETURN { title: m.title, embedding_dimensions: LENGTH(emb.embedding) }
> ```

Reviewer: ```suggestion
```

please do not promote DOCUMENT prominently.
The alternative below is better as it can be optimized.
DOCUMENT is a black box, and does not profit from CoLocated data.
Furthermore DOCUMENT will add the Collection access at runtime of the query.

For each movie, `LET emb = ...` fetches its matching embedding document, and the

> `RETURN` combines fields from both. The general, collection-agnostic form of a

Reviewer: It is exactly not collection-agnostic.

DOCUMENT is collection agnostic and will pick up whatever it finds.

FOR is bound to a specific collection

join is a nested `FOR` with a `FILTER` that relates the two sides:

```aql
FOR m IN Movies
  FILTER m.vote_average >= 8
  LIMIT 5
  FOR e IN MovieEmbeddings
    FILTER e._key == m._key
    RETURN { title: m.title, embedding_dimensions: LENGTH(e.embedding) }
```

Both queries produce the same result. The `DOCUMENT()` function is typically a
bit faster when looking up a few individual documents but the `FOR` loop
is better when looking up many documents.

This kind of join follows a shared *value*. When relationships are modeled as
edges instead, you follow them with a graph traversal — that is the next step.

## Step 10: Traverse the graph

A graph traversal follows edges from a starting document. The core syntax names
three variables (the `node` reached, the `edge` used, and the `path` so far), a
depth range, a direction, and the edge collection to walk. Let's switch to Toy
Story (`Movies/862`) and list its genres by following `HasGenre` edges outbound
from the movie to the genre nodes:

```aql

> FOR node, edge, path IN 1..1 OUTBOUND "Movies/862" HasGenre

Reviewer: Please start with a Pickup by name not by ID.

FOR m in Movies
  FILTER m.title == "Toy Story"

  RETURN node.name
```

Direction matters. `ActedIn` edges point from a person **to** a movie, so to find
a movie's cast you traverse `INBOUND` — from the movie back to the people:

```aql

> FOR person IN 1..1 INBOUND "Movies/862" ActedIn

Reviewer: Here you suddenly omit 2 variables.

Which is correct we know, but could be confusing.

  RETURN person.name
```

Increase the depth to answer connected questions. A two-step traversal from a
movie to its actors and on to *their* other movies finds films that share cast
with Toy Story:

```aql
FOR person IN 1..1 INBOUND "Movies/862" ActedIn
  FOR other IN 1..1 OUTBOUND person ActedIn
    FILTER other._id != "Movies/862"
    RETURN DISTINCT other.title
```

Due to the topology of the example dataset (Movie → Person → Movie), this can be
simplified to a single traversal with a depth of 2 and following inbound as well
as outbound edges using `ANY` as the direction:

```aql
FOR movie IN 2..2 ANY "Movies/862" ActedIn

>     FILTER movie._id != "Movies/862"

Reviewer: ```suggestion
```

This filter is not necessary here.
Worth mentioning, that this 2 hop traversal will not  use the same edge twice on a path.

Author: Yes, as long as there isn't another edge between the two - need to check if there's something like this in the dataset.

    RETURN DISTINCT movie.title
```

You can vary the depth range (`1..2`), the direction (`INBOUND`, `OUTBOUND`, or
`ANY`), and how strictly nodes and paths must be unique, to shape what a
traversal explores. Instead of naming edge collections one by one, you can also
traverse the whole named graph at once with `GRAPH "MoviesGraph"`, which walks
every edge collection the graph defines:

```aql

> FOR node IN 1..1 OUTBOUND "Movies/862" GRAPH "MoviesGraph"
>   RETURN node

Reviewer: ```suggestion
FOR node IN 1..1 OUTBOUND "Movies/862" GRAPH "MoviesGraph"
  RETURN {
    type: PARSE_COLLECTION(node._id),
    name: node.name
  }
```

This formats the response nicer to read.
Otherwise you get a mostly empty table.

```

## Step 11: Find similar movies with vector search

So far we matched exact values (title, rating) and explicit relationships
(edges). Vector search finds movies by **meaning**: each movie's `overview` text
was turned into a 384-dimensional embedding in `MovieEmbeddings`, and similar
plots end up close together in that vector space.

First, create the index once. In the web interface, open the `MovieEmbeddings`
collection, go to its **Indexes** tab, and add a **vector index** on the
`embedding` field with metric `cosine` and dimension `384` (the output size of

> the `all-MiniLM-L6-v2` model that produced the embeddings).

Reviewer: We also need to mention the Number of Lists (nLists).
It is `15 * sqrt(N)` in this case is `3.200`

Reviewer: When creating the Index I do get:
"Something went wrong in index creation: Index not found"

{{< info >}}
Vector indexes require the deployment to be started with the `--vector-index`
startup option. On a platform from [Evaluate locally](evaluate-locally.md) this
is already set; on your own deployment, enable it before creating the index.
{{< /info >}}

Now run a "more like this" query. Take Toy Story's own embedding as the reference
point and ask the index for its nearest neighbors:

```aql

> LET query = DOCUMENT("MovieEmbeddings/862").embedding  // Toy Story
> FOR e IN MovieEmbeddings
>   SORT APPROX_NEAR_COSINE(e.embedding, query) DESC
>   LIMIT 6
>   RETURN DOCUMENT("Movies", e._key).title

Reviewer: ```suggestion
LET query = DOCUMENT("MovieEmbeddings/862").embedding  // Toy Story

FOR e IN MovieEmbeddings
  SORT APPROX_NEAR_COSINE(e.embedding, query) DESC
  LIMIT 6
  FOR m IN Movies
    FILTER m._key == e._key
    RETURN {
      title: m.title
    }
```

This formats it a bit nicer

```

The results are the movies whose overviews are **semantically closest** to Toy
Story — expect other animated, family, and toy-themed films — even when they
share no genre, keyword, or cast edge in the graph. This is similarity by
meaning, not by exact attribute match.

The moving parts: `APPROX_NEAR_COSINE` returns a cosine similarity (`1` = most
similar) computed against the vector index, `SORT ... DESC` puts the closest
matches first, and `LIMIT` caps how many the index returns. The reference movie
itself is always the top hit (similarity ≈ 1), which is why we ask for 6 to get
5 genuinely *other* recommendations.

> 

Reviewer: This stops too early.

This is the point where our product can actually shine. Not the stuff above.

The combination of everything people have learned before.

Therefor I would suggest to start a next step:

Real life Use-cases.

Give me a movie with at least 5 same actors then Toy Story sorted by best semantic match:

FOR t In Movies
  FILTER t.title == "Toy Story"
  FOR embeddingDoc IN MovieEmbeddings
    FILTER embeddingDoc._key == t._key
    LET embedding = embeddingDoc.embedding // Toy Story embedding
    LET moviesThatShare5Actors = (
      FOR otherMovie in 2 ANY t ActedIn
        COLLECT id = otherMovie._key, title = otherMovie.title WITH COUNT INTO c
        FILTER c >= 5 // Share at least 5 actors
        RETURN {id, title}
    )
    FOR otherMovie IN moviesThatShare5Actors
      FOR otherEmbeddingDoc IN MovieEmbeddings
        FILTER otherEmbeddingDoc._key == otherMovie.id
        LET similarity = COSINE_SIMILARITY(embedding, otherEmbeddingDoc.embedding) 
        SORT similarity DESC
        LIMIT 3
        RETURN {title: otherMovie.title, similarity}

Maybe we can find one or two more real life examples here.

Happy to brainstorm.
